### ISSUE TITLE
Excessive Core Data notifications when syncing with relationships

### ISSUE THREAD

**Issue by Jonge** [3lvis/Sync#500]
Hey,

I use `NSFetchedResultsController` with Sync.

I found that when you have JSON objects with relationships and call Sync, Core Data notifications (and therefore also `NSFetchedResultsControllerDelegate` methods) are called way too often -- after syncing every relationship object. This results in:
- janky UI, because the rows are inserted/updated one by one;
- performance issues, because the UITableView needs to do animations one by one instead of a batch.

For a single Sync operation, I would expect a single merge/save notification. `NSFetchedResultsControllerDelegate` methods should be called like this:

```
controllerWillChangeContent
controller(_, didChange anObject: ...)
controller(_, didChange anObject: ...)
controller(_, didChange anObject: ...)
controllerDidChangeContent
```

But instead, I get this:

```
controllerWillChangeContent
controller(_, didChange anObject: ...)
controllerDidChangeContent
controllerWillChangeContent
controller(_, didChange anObject: ...)
controllerDidChangeContent
controllerWillChangeContent
controller(_, didChange anObject: ...)
controllerDidChangeContent
```

I tried a quick hack, and while it works for my project, it breaks some tests, because the relationships are not saved immediately, which breaks some use-cases: https://github.com/Jonge/Sync/tree/bug-excessive-notifications

**Comment by batjo** [comment]
I noticed the same problems, didn't know it was because of the many Notifications. Would be nice if there's a solution that passes all the tests

**Comment by 3lvis** [comment]
This is another alternative to solve performance issues. https://github.com/3lvis/performance

**Comment by 3lvis** [comment]
Difficult part here is that we don't handle notifications that's done by Core Data, if you have any other suggestions I'll be happy to look at it. <3

**Comment by Jonge** [comment]
I understand. Well the notifications are sent every time you `save()` the main context. In theory, we should merge to the main context and `save()` it only after syncing all objects.

However, DataStack uses notifications and merges to main context on every background context save, and I don't see an easy way to change that currently.

**Comment by 3lvis** [comment]
The problem with not saving is the relationships, you saw how the tests break when you stop saving...

**Comment by Jonge** [comment]
Yes, I know. :(

Sorry I only have theoretical knowledge here, but it should be possible to have an "intermediate" context, fetch and update the relationships there, then merge into main context. But again, doesn't seem easy to do with the notifications merging it into main context every time.

I might try looking at this in-depth, mainly how the relationships are fetched and updated.

**Comment by 3lvis** [comment]
The main challenge right now with not merging is many-to-many. If you don't merge when you get to the many-to-many operation then Sync doesn't know about the new relationships so it doesn't work. In theory we could potentially see if there's a many-to-many operation involved and if not then avoid merging until the very end.
